src/main/scala/org/broadinstitute/hail/expr/AST.scala

```diff
@@ -1622,6 +1623,12 @@ case class Let(posn: Position, bindings: Array[(String, AST)], body: AST) extend
 }
 
 case class BinaryOp(posn: Position, lhs: AST, operation: String, rhs: AST) extends AST(posn, lhs, rhs) {
+  def liftBinaryWithCoerce[T,U,V](coerce: T => U)(op: (U, U) => V): (T, T) => V =
+    (l, r) => op(coerce(l), coerce(r))
+
+  def threadNulls(f: (Any, Any) => Any): (Any, Any) => Any =
```

Comment:
I was thinking:

```scala
  def evalNumeric[T1, T2](g: (T1, T2) => Any): (() => Any, () => Any) => Any
```

which could be used to implement `evalComposeNumeric`. We could also expose a non-thunked version.